openassessment/staffgrader/models/submission_lock.py

```diff
@@ -0,0 +1,83 @@
+"""
+Models for locking Submissions for exclusive grading.
+Part of Enhanced Staff Grader (ESG).
+"""
+from django.db import models
+from django.utils.timezone import now
+from django.utils.translation import ugettext as _
+
+from openassessment.assessment.models.staff import StaffWorkflow
+from openassessment.staffgrader.errors.submission_lock import SubmissionLockContestedError
+
+
+class SubmissionGradingLock(models.Model):
+    """
+    Internal model for locking a submission for exclusive grading
+    """
+    TIMEOUT = StaffWorkflow.TIME_LIMIT
+
+    # NOTE - submission_uuid can refer to either the team or individual submission
+    submission_uuid = models.CharField(max_length=128, db_index=True, unique=True)
+    owner_id = models.CharField(max_length=40, db_index=True)
+    created_at = models.DateTimeField(db_index=True)
+
+    class Meta:
+        app_label = "staffgrader"
+        ordering = ["created_at", "id"]
+
+    @property
+    def is_active(self):
+        """
+        Check if lock is still active (has not timed out)
+        """
+        return now() < self.created_at + self.TIMEOUT
+
+    @classmethod
+    def get_submission_lock(cls, submission_uuid):
+        """
+        Get info about a submission grading lock
+
+        Returns: SubmissionGradingLock info or None
+        """
+        return cls.objects.filter(submission_uuid=submission_uuid).first()
+
+    @classmethod
+    def claim_submission_lock(cls, submission_uuid, user_id):
+        """
+        Try to claim a submission grading lock
+
+        Returns: SubmissionGradingLock or raises Error
+        """
+        # See if there's already a lock
+        current_lock = cls.get_submission_lock(submission_uuid)
+
+        # If there's already an active lock, raise an error
+        if current_lock and current_lock.is_active:
+            raise SubmissionLockContestedError(_("Submission already locked"))
+
+        # Otherwise, create a new lock
+        new_lock, created = cls.objects.update_or_create(  # pylint: disable=unused-variable
+            owner_id=user_id,
+            created_at=now(),
+            submission_uuid=submission_uuid,
+        )
+
+        return new_lock
+
+    @classmethod
+    def clear_submission_lock(cls, submission_uuid, user_id):
+        """
+        Clear an existing lock. Locks can only be cleared by the lock owner
+
+        Returns: None or raises Error
+        """
+        # Get the current lock, assuming it exists
+        current_lock = cls.get_submission_lock(submission_uuid)
+        if not current_lock:
+            return
+
+        # Only the owner can clear the lock
+        if current_lock.owner_id != user_id:
+            raise SubmissionLockContestedError(_("Unable to clear submission lock"))
+
+        current_lock.delete()
```

Comment:
mmm, I think we shouldn't delete these models. I thought the reason this is a new model was for traceablility? If we aren't keeping these around, there's really no point of adding this model when we already have the scorer_id and graded_started_at on the staff workflow

Reply:
I don't really follow what you mean by that. Do we want to keep a history of who locked what or no? What benefit does this model give us over what we have already then?

Reply:
We don't really need a history. What this buys us over the previous system is allowing us to add locks for regrading. Otherwise, we'd have to manage timestamps on completed staff workflows which would be gross. This also allows us future ability to expand how we do lock management, like tying a lock to a session instead of a user ID.